Flaky test `RaftSnapshotReplicationFailureHandlingTest.shouldResentSnapshotIfFirstChunkTimedOut`

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? just [once](https://github.com/camunda/zeebe/actions/runs/8375596556/job/22933276898)
- Does it block your work?
- Do we suspect that it is a real failure?

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
Warning:  Flakes: 
Warning:  io.atomix.raft.RaftSnapshotReplicationFailureHandlingTest.shouldResentSnapshotIfFirstChunkTimedOut
Error:    Run 1: RaftSnapshotReplicationFailureHandlingTest.shouldResentSnapshotIfFirstChunkTimedOut:109 » ConditionTimeout Condition with alias 'Snapshot chunk-0 is resend' didn't complete within 10 seconds because assertion condition defined as a Lambda expression in io.atomix.raft.RaftSnapshotReplicationFailureHandlingTest [Should resent snapshot chunk] 
expected: 2
 but was: 3.
Error:    Run 2: RaftSnapshotReplicationFailureHandlingTest.shouldResentSnapshotIfFirstChunkTimedOut:109 » ConditionTimeout Condition with alias 'Snapshot chunk-0 is resend' didn't complete within 10 seconds because assertion condition defined as a Lambda expression in io.atomix.raft.RaftSnapshotReplicationFailureHandlingTest [Should resent snapshot chunk] 
expected: 2
 but was: 3.
[INFO]   Run 3: PASS
</pre>
</details>

**Hypotheses**

> List any hypotheses if you have one; can be ommitted

**Logs**

> If possible, provide more context here, e.g. standard output logs, link to build, etc.

Full test results: [Google Drive](https://drive.google.com/file/d/1D12F1Qwm7UAqiROVTBr36WpI_jm6-N8G/view?usp=drive_link)

<details><summary>Logs</summary>
<pre>
Error:  Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 11.59 s <<< FAILURE! -- in io.atomix.raft.RaftSnapshotReplicationFailureHandlingTest
Error:  io.atomix.raft.RaftSnapshotReplicationFailureHandlingTest.shouldResentSnapshotIfFirstChunkTimedOut -- Time elapsed: 11.58 s <<< ERROR!
org.awaitility.core.ConditionTimeoutException: 
Condition with alias 'Snapshot chunk-0 is resend' didn't complete within 10 seconds because assertion condition defined as a Lambda expression in io.atomix.raft.RaftSnapshotReplicationFailureHandlingTest [Should resent snapshot chunk] 
expected: 2
 but was: 3.
	at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:167)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:1006)
	at org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:790)
	at io.atomix.raft.RaftSnapshotReplicationFailureHandlingTest.shouldResentSnapshotIfFirstChunkTimedOut(RaftSnapshotReplicationFailureHandlingTest.java:109)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
Caused by: org.opentest4j.AssertionFailedError: [Should resent snapshot chunk] 
expected: 2
 but was: 3
	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
	at io.atomix.raft.RaftSnapshotReplicationFailureHandlingTest.lambda$shouldResentSnapshotIfFirstChunkTimedOut$1(RaftSnapshotReplicationFailureHandlingTest.java:113)
	at org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:53)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:248)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:235)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
</pre>
</details>

--- a/zeebe/atomix/cluster/src/test/java/io/atomix/raft/RaftSnapshotReplicationFailureHandlingTest.java
+++ b/zeebe/atomix/cluster/src/test/java/io/atomix/raft/RaftSnapshotReplicationFailureHandlingTest.java
@@ -95,7 +95,8 @@ public class RaftSnapshotReplicationFailureHandlingTest {
     final int numberOfChunks = 1;
     disconnectFollowerAndTakeSnapshot(numberOfChunks);
 
-    leaderProtocol.interceptResponse(InstallResponse.class, new TimingOutInterceptor(1));
+    final TimingOutInterceptor responseInterceptor = new TimingOutInterceptor(1);
+    leaderProtocol.interceptResponse(InstallResponse.class, responseInterceptor);
 
     // when
     reconnectFollowerAndAwaitSnapshot();
@@ -108,9 +109,13 @@ public class RaftSnapshotReplicationFailureHandlingTest {
     Awaitility.await("Snapshot chunk-0 is resend")
         .untilAsserted(
             () ->
-                assertThat(totalInstallRequest.get())
+                assertThat(responseInterceptor.getCount())
                     .describedAs("Should resent snapshot chunk")
-                    .isEqualTo(numberOfChunks + 1));
+                    // Before follower reconnects, sometimes leader sends an InstallRequest which
+                    // ends up in connect exception. So comparing number of requests results in
+                    // flaky test. Instead, compare there were two responses. First one is timeout
+                    // and second is success.
+                    .isEqualTo(2));
   }
 
   private void reconnectFollowerAndAwaitSnapshot() throws InterruptedException {
@@ -150,6 +155,10 @@ public class RaftSnapshotReplicationFailureHandlingTest {
       } else {
         return CompletableFuture.completedFuture(installResponse);
       }
+    }
+
+    int getCount() {
+      return count;
     }
   }
 

